### DISCUSSION TITLE
Unable to update formfield controlled by useFieldArray from another component using setValue

### DISCUSSION THREAD

**Discussion by rvanpraet** [react-hook-form/react-hook-form#5615] [Q&A]
**Describe the bug**
In my use case, I'm using a form with multible tabs, one of the tabs contains a useFieldArray which is part a larger form. Using useFormContext to pass down control. An action in another tab which is also part of the form should be able to update the form field controlled by useFieldArray. I'm trying to use setValue for this, I can see that the value is being set correctly by calling getValues in the FieldArray component.

However, the new values do not get rendered and 'fields' variable still logs with only two values.

**Relevant files:**
- TabInfo.tsx --> Button onClick function generates new entries for the 'persons' form field.
- FieldArray.tsx --> Logs values and where the change is expected to be seen.

**To Reproduce**

1. Navigate to 'Employees' tab, check the console logs. Both 'fields' and 'getValues' log the same values.
2. Go back to 'Info' tab, click on 'Update Fieldarray'
3. Go to 'Employees' tab, the log will show getValues with the same fieldname that useFieldArray is using to have 3 values. fields variable logs with only 2 values. Only two values get rendered.
4. Press 'DEL' on one of the field arrays, the app will crash.

**Codesandbox link (Required)**
https://codesandbox.io/s/rhf-updating-ufa-outside-p88ge

**Expected behavior**
I would like to be able to update the form field that is registered to useFieldArray from another component using setValue.

**Desktop (please complete the following information):**
- OS: Windows 10
- Browser: Firefox
- Version 89.0

**Additional context**

**Comment by bluebill1049** [comment]
thanks for the detailed issue report, but please provide an isolated codesandbox to help me with the investigation.

**Comment by rvanpraet** [comment]
> 
> 
> thanks for the detailed issue report, but please provide an isolated codesandbox to help me with the investigation.

Hi Bill, thanks for the swift reply. I have made a stripped down version of the CodeSandbox which doesn't use Tabs and here the described bug doesn't seem to happen:

https://codesandbox.io/s/rhf-updating-ufa-minimal-6tbcw?file=/src/App.tsx

I'm not entirely sure how I can isolate the original issue, it seems to be related to the form being tabbed (perhaps useFieldArray not rendering together with the button making the change). 

I slightly edited the original Codesandbox, removed unrelated fields on the first Tab. The rest of the code is standard Material UI Tabs and Controller wrappers around Material UI TextField and Autocomplete. Hope that even though the codesandbox is a bit extensive, it displays the issue well and that you can have a look at it!

**Comment by bluebill1049** [comment]
Thanks @rvanpraet for the updated CSB, can you share some steps as well? i will take a look at it tonight.

**Comment by bluebill1049** [comment]
<img width="728" alt="Screen Shot 2021-06-15 at 6 25 31 pm" src="https://user-images.githubusercontent.com/10513364/122019405-135eb980-ce07-11eb-917b-3e7f19af65db.png">

is it this part?

**Comment by rvanpraet** [comment]
Indeed this part. In the CSB without tabs this functionality works, in the original CSB (with tabs) provided in my first post I can't seem to get it to work.

**Reply by bluebill1049** [reply]
I think I get the problem now because useFieldArray is not been executed and hence the useFieldArray is really there but you are trying to set it.

## potential solution:

move useFieldArray outside of the tab view, so it will get ruined even the tab view is not presented. let me know if this solve your problem.

**Reply by rvanpraet** [reply]
Hi Bill, sorry for the late reply. This indeed does seem to work, if I call useFieldArray and store it's methods in a variable and pass it down as props to the tab that needs it, the field arrays updates correctly:

https://codesandbox.io/s/rhf-updating-ufa-outside-p88ge?file=/src/App.tsx

However, for my main application this way of passing the useFieldArray methods would be cumbersome to say the least. I'm using a generic approach and have splitted useForm and useFieldArray in their own wrapper components so I can use them in a modular way, connected with useFormContext. 

This potential solution would force me to declare every useFieldArray in the same wrapper component as useForm, causing issues if the form has more than 1 fieldArray. I could have to find a way to iterate over fieldnames and have the wrapper component return seperate useFieldArray methods for each one, to then pass them down to the correct tab.

Ideally I would want to use the useFieldArray wrapper in a tab which lives under the FormContext and can be registered/updated via the fieldname without it having to render or be declared first. Basically this is also how the rest of the form fields work, so it should be possible to do it for useFieldArray?

**Reply by charleshimmer** [reply]
We are running into an similar issue. We desire for `setValue` to be able to update a nested `FieldArray`'s value where hosting the methods for that particular instance of fieldArray would not be really feasible. I think technically we could hoist those methods onto a context somewhere, but I think that would start to get gross and pretty cumbersome.
